### PR TITLE
[SPARK-43226] Define extractors for file-constant metadata

### PR DESCRIPTION
**Pull request by ryan-johnson-databricks**
### What changes were proposed in this pull request?

File-source constant metadata columns are often derived indirectly from file-level metadata values rather than exposing those values directly. Add support for metadata extractors, so that we can express such columns in a generic way.

### Why are the changes needed?

Allows to express the existing file-source metadata columns in a generic way (previously hard-wired), and also allows to lazily derive expensive metadata values only if the query actually selects them.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

New unit test. Plus, existing file-source metadata unit tests pass.

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileFormat.scala [hunk 1]
```diff
@@ -241,47 +256,74 @@ object FileFormat {
...
+   * file's metadata map, returning null if not found.
+   *
+   * Raw values (including null) are automatically converted to literals as a courtesy.
+   */
+  def getFileConstantMetadataColumnValue(
+      name: String,
+      file: PartitionedFile,
+      metadataExtractors: Map[String, PartitionedFile => Any]): Literal = {
+    val extractor = metadataExtractors.get(name).getOrElse {
+      (_: PartitionedFile).otherConstantMetadataColumnValues.get(name).orNull
```
Comment (johanl-db) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileFormat.scala:287]: Note: the syntax
```
(_: PartitionedFile).otherConstantMetadataColumnValues.get(name).orNull
```
to define the default extractor threw me off at first, the following would be easier to understand:
```
pf: PartitionedFile => pf.otherConstantMetadataColumnValues.get(name).orNull
```

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileFormat.scala [hunk 2]
```diff
@@ -203,6 +203,21 @@ trait FileFormat {
...
+
+  /**
+   * The extractors to use when deriving file-constant metadata columns for this file format.
+   *
+   * A scanner must derive each file-constant metadata field's value from each [[PartitionedFile]]
+   * it processes. By default, the value is obtained by a direct lookup of the column's name on
+   * [[PartitionedFile.otherConstantMetadataColumnValues]] (see
+   * [[FileFormat.getFileConstantMetadataColumnValue]]). However, implementations can override this
+   * method in order to provide more sophisticated lazy extractors (e.g. in case the column value is
+   * complicated or expensive to compute).
```
Comment (johanl-db) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileFormat.scala:215]: Someone adding a new constant metadata column in the future will need to choose between the two alternatives to provide values, this choice could be described more explicitly here:

- extractor: anytime the value can be directly derived from `PartitionedFile` information.
- `PartitionedFile.otherConstantMetadataColumnValues`: all other cases, e.g. the value is arbitrary or is coming from some other place

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileFormat.scala (hunk 1 above)
Comment (ryan-johnson-databricks) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileFormat.scala:287]: Yup. I already came to the same conclusion and fixed it on the OSS side. This side will get fixed when I rebase on the OSS sync (if not sooner).

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileFormat.scala (hunk 2 above)
Comment (ryan-johnson-databricks) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileFormat.scala:215]: I thought I _did_ describe it explicitly:
1. If you provide an extractor, the extractor has access to all state in the `PartitionedFile` (including the column value map) and can do any computations it needs to.
2. Otherwise, the column's value is fetched from the column value map.

Was it not clear [enough] in the comment that the extractor has access to the entire `PartitionedFile` ?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileFormat.scala (hunk 1 above)
Comment (ryan-johnson-databricks) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileFormat.scala:287]: Agreed. I think it's already fixed locally but I didn't push yet in case there were other comments.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileFormat.scala (hunk 2 above)
Comment (ryan-johnson-databricks) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileFormat.scala]: Reworded, PTAL

Comment (johanl-db) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileFormat.scala]: The way I read it is that using `otherConstantMetadataColumnValues` is the default and an extractor should only be used for more advanced use cases. I think we rather want to encourage using extractors whenever possible and only fallback to manually providing the values when necessary.

Comment (ryan-johnson-databricks) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileFormat.scala]: That wording was intentional... the simple way is usually the right way and thus should be the default. If something special is going on, I'd prefer it to be highly visible. 

Put another way -- if the column _can_ be represented simply by pulling the value from the map, why should we encourage the use of an extractor? 

An alternative would be to require all constant metadata columns to provide extractors, and define a default extractor that just goes to the map... but this seems like extra boilerplate for no real benefit? 

(I originally wanted to just embed the extractor in the attribute definition directly, but attribute/struct metadata cannot store code and IMO that's a wise design decision we shouldn't change)

**Comment by ryan-johnson-databricks** [comment]
FYI the [tests that failed](https://github.com/ryan-johnson-databricks/spark/actions/runs/4765599580/jobs/8471553389) are broken upstream -- they also fail on the version of spark/master this PR is currently based on.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileFormat.scala [hunk 3]
```diff
@@ -203,6 +203,21 @@ trait FileFormat {
...
+   * By default, the value of a file-constant metadata column is obtained by looking up the column's
+   * name in the file's metadata column value map. However, implementations can override this method
+   * in order to provide an extractor that has access to the entire [[PartitionedFile]] when
+   * deriving the column's value.
+   *
+   * NOTE: Extractors are lazy, invoked only if the query actually selects their column at runtime.
+   *
+   * See also [[FileFormat.getFileConstantMetadataColumnValue]].
+   */
+  def fileConstantMetadataExtractors: Map[String, PartitionedFile => Any] =
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileFormat.scala:219]: I have 2 questions about the API design:
1. File source implementations now need to define constant metadata columns in both `metadataSchemaFields` and `fileConstantMetadataExtractors`. Shall we have a single API to define constant metadata columns? The API should define both column name, type, and extractor.
2. Should the extractor return a `Literal` instead of `Any`?

Comment (ryan-johnson-databricks) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileFormat.scala:219]: > Shall we have a single API to define constant metadata columns?

Maybe? I couldn't figure out a good way to do it:

> I originally wanted to just embed the extractor in the attribute definition directly, but attribute/struct metadata cannot store code and IMO that's a wise design decision we shouldn't change

I guess we could store pairs or a triple, but (a) the extractors are optional (we'd be forcing every column to either use one or opt out); and (b) the uses for the attribute/struct definitions are rather separated from the uses of the extractors (result: a lot of `x.map` and `x.filter` type calls clutter up the code)

> Should the extractor return a `Literal` instead of `Any`?

I thought about it? But:
* it complicates the definitions of all extractors with boilerplate, which IMO indicates a flawed design
* we anyway have to wrap it in a literal again to handle the null case (or require every extractor to think about nulls). 
* Currently, `getFileConstantMetadataColumnValue` takes care of converting to literal, and that's what everyone uses (nobody calls the extractor directly).
* It would be non-symmetric with the default case, that pulls `Any` from the column value map.

Note that nothing prevents an extractor from returning a literal, because `Literal.apply` will detect and return it unchanged.

Overall, I worry the above changes would only give the code a bigger surface area with little/no gain in safety or functionality?

Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileFormat.scala:219]: Can we have 2 APIs?
```
def constantMetadataCols: Seq[(StructField, PartitionedFile => Any)] = ...

def generatedMetadataCols: Seq[StructField]
```

Comment (ryan-johnson-databricks) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileFormat.scala:219]: My earlier response assumed you meant two APIs, and was pointing out that it would cause headaches for both user and implementation because most metadata columns don't need extractors, and the code that uses extractors is different from the code that uses the attributes.

Comment (ryan-johnson-databricks) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileFormat.scala:219]: Meanwhile, what "safety" issue are we even responding to here? Right now, the behavior of a file-constant metadata column is perfectly predictable (and consistent with its behavior before this PR) -- _unless_ the user opts in to using an extractor by overriding the method to provide an extractor for it.

Comment (ryan-johnson-databricks) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileFormat.scala:219]: Today's API is two pieces:
```scala
/** All fields of the _metadata struct */
def metadataSchemaFields: Seq[StructField]
/** Custom extractors for the subset of file-constant metadata columns that actually need them */
def fileConstantMetadataExtractors: Map[String, PartitionedFile => Any]
```
Splitting the API would produce instead (at a minimum):
```scala
/** Extractor for file-constant metadata columns that just pull the value from the metadata map */
val DEFAULT_EXTRACTOR: PartitionedFile => Any
/** Defines the file-constant fields of the _metadata struct */
def constantMetadataCols: Seq[(StructField, PartitionedFile => Any)]
/** Defines the file-generated fields of the _metadata struct */
def generatedMetadataCols: Seq[StructField]
```
The only safety issue I thought of so far, is that it might be possible to get the name mapping wrong between columns and their extractors. But if that happened even the most basic unit test for the new column would fail. Meanwhile, that mapping is anyway needed internally because the list of metadata columns is named attributes or even column names, so the pairing proposed above would anyway have to be reconstructed.

Comment (ryan-johnson-databricks) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileFormat.scala:219]: I just realized a potential difference of underlying assumptions that could be making this discussion more difficult:

The reviewer comments so far are focused on what the changes in this PR look like. However, this PR isn't representative of the experience somebody would have while trying to define custom metadata columns for their file format. By definition all of the standard columns `FileFormat` natively provides are _NOT_ custom and come directly from the fields of `PartitionedFile` itself, rather than indirectly from the metadata values map which users have control over.

My thought process for designing the API was more focused on ease of use for the former case (custom file formats), since the latter (base file format) is a one-time effort within in this PR.

Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileFormat.scala:219]: My preference is to have a simple and clean framework to generate constant metadata columns. I think the framework is a bit complicated right now:
1. FileFormat implementations can define a few constant metadata columns (name and data type).
2. The constant metadata columns should either match a built-in name in `FileScanRDD`, or be filled by a custom `FileIndex`. With this PR, there is no built-in name. It's either filled by `FileIndex` or has an extractor.
3. With this PR, FileFormat implementations can provide extractors for certain constant metadata columns.

Basically, we get the value of a constant metadata column from two maps: one map is filled by `FileIndex` and contains the value directly. The other map is filled by the `FileFormat` implementation and contains extractors.

It will be great if we can replace the first map with the second one. But if we can't, I'm fine with the current API as we can't avoid name lookup anyway. Can we add more comments in the new API, saying that it may only handle part of the constant metadata columns? The rest of them should be handled by a custom `FileIndex`.

Comment (ryan-johnson-databricks) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileFormat.scala:219]: I made a dummy PR https://github.com/apache/spark/pull/40930 to show what the other approach actually looks like. I still think it is uglier code for everyone involved, with no benefit to safety, but at least this way we have something concrete to compare.

The change for the other PR vs master is:
```
 sql/core/src/main/scala/org/apache/spark/sql/execution/DataSourceScanExec.scala                              |  33 +++++++++++++---------------
 sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileFormat.scala                          | 145 ++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++--------------------------------------
 sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileScanRDD.scala                         |  70 +++++++++++++++--------------------------------------------
 sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/PartitioningAwareFileIndex.scala          |  10 ++++-----
 sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFileFormat.scala           |   4 ++--
 sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/FileSourceCustomMetadataStructSuite.scala |  71 ++++++++++++++++++++++++++++++++++++++++++++++++------------
 6 files changed, 196 insertions(+), 137 deletions(-)
```

The change of this PR vs. master touches fewer files and is ~50LoC smaller:
```
 sql/core/src/main/scala/org/apache/spark/sql/execution/DataSourceScanExec.scala                              |  33 +++++++++++++++------------------
 sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileFormat.scala                          | 102 ++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++------------------------------
 sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileScanRDD.scala                         |  70 +++++++++++++++++-----------------------------------------------------
 sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/PartitioningAwareFileIndex.scala          |  10 +++++-----
 sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/FileSourceCustomMetadataStructSuite.scala |  35 ++++++++++++++++++++++++++++++++++-
 5 files changed, 143 insertions(+), 107 deletions(-)
```

The other PR vs. this one is changes 3 files and ~100LoC:
```
 sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileFormat.scala                          | 99 ++++++++++++++++++++++++++++++++++++++++++++++++++++++++-------------------------------------------
 sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFileFormat.scala           |  4 ++--
 sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/FileSourceCustomMetadataStructSuite.scala | 70 ++++++++++++++++++++++++++++++++++++++++------------------------------
 3 files changed, 98 insertions(+), 75 deletions(-)
```

Comment (ryan-johnson-databricks) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileFormat.scala:219]: > Basically, we get the value of a constant metadata column from two maps: one map is filled by FileIndex and contains the value directly. The other map is filled by the FileFormat implementation and contains extractors.

Nit: The extractors don't "contain" the value at all. They just tell how to extract a value from the `PartitionedFile` that the `FileIndex` provides. I don't know a cleaner way to do this, given that the extraction policy should probably be consistent across all files, rather than re-defined by each file? (prior to this PR, the "policy" was hard-wired as "look it up in the map"; now it's allowed to also be "do arbitrary computation on `PartitionedFile`, if the user prefers)

Comment (ryan-johnson-databricks) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileFormat.scala:219]: > Can we add more comments in the new API, saying that it may only handle part of the constant metadata columns? The rest of them should be handled by a custom FileIndex.

There was already one comment to that effect in the PR that added custom metadata column support:
https://github.com/apache/spark/pull/40677/files#diff-2d7ec6604810e1be58ca7852ec389853c143b684894f2390ffe07f96701d2219R29

And the unit tests also are a pretty clear demonstration. 

Good point that the comments are buried pretty deep tho -- I'll try to add/expand them in FileFormat.scala as well.

Comment (ryan-johnson-databricks) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileFormat.scala:223]: doc comments updated, PTAL

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileFormat.scala [hunk 4]
```diff
@@ -241,47 +260,74 @@ object FileFormat {
     FileSourceConstantMetadataStructField(FILE_BLOCK_LENGTH, LongType, nullable = false),
     FileSourceConstantMetadataStructField(FILE_MODIFICATION_TIME, TimestampType, nullable = false))
 
+  /**
+   * All [[BASE_METADATA_FIELDS]] require custom extractors because they are derived directly from
+   * fields of the [[PartitionedFile]], and do have entries in the file's metadata map.
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileFormat.scala:265]:
```suggestion
   * fields of the [[PartitionedFile]], and don't have entries in the file's metadata map.
```

Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileFormat.scala:265]: @ryan-johnson-databricks do I understand this correctly?

**Review by cloud-fan — APPROVED** [review]

**Comment by cloud-fan** [comment]
thanks, merging to master!

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileFormat.scala (hunk 4 above)
Comment (ryan-johnson-databricks) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileFormat.scala:265]: Erm, yes... I think I was trying to type "do not" but didn't quite make it.
